File: scanner/tests/fixtures/vulnerable_app.swift
import Foundation

class DatabaseService {
    var connectionString: String

    init(connectionString: String) {
        self.connectionString = connectionString
    }

    func getUser(userId: String) -> String {
        let query = "SELECT * FROM users WHERE id = '\(userId)'"
        return executeQuery(query: query)
    }

    func executeQuery(query: String) -> String {
        return query
    }
}

class CommandRunner {
    func runCommand(input: String) {
        let process = Process()
        process.executableURL = URL(fileURLWithPath: "/bin/sh")
        process.arguments = ["-c", input]
        try? process.run()
    }
}

func handleUserInput(input: String) -> String {
    let db = DatabaseService(connectionString: "sqlite:///mydb")
    return db.getUser(userId: input)
}

func executeShell(command: String) {
    let runner = CommandRunner()
    runner.runCommand(input: command)
}

File: scanner/tests/fixtures/vulnerable_app.rs
use std::process::Command;
use std::env;

struct Database {
    conn_str: String,
}

impl Database {
    fn new(conn_str: &str) -> Self {
        Database { conn_str: conn_str.to_string() }
    }

    fn query_user(&self, user_id: &str) -> String {
        let query = format!("SELECT * FROM users WHERE id = '{}'", user_id);
        self.execute(&query)
    }

    fn execute(&self, query: &str) -> String {
        query.to_string()
    }
}

fn handle_request(user_input: &str) -> String {
    let db = Database::new("postgres://localhost/mydb");
    db.query_user(user_input)
}

fn run_command(cmd: &str) {
    let output = Command::new("sh")
        .arg("-c")
        .arg(cmd)
        .output()
        .expect("failed");
    println!("{}", String::from_utf8_lossy(&output.stdout));
}

fn get_env_data() -> String {
    let val = env::var("USER_INPUT").unwrap_or_default();
    run_command(&val);
    val
}
